### ISSUE TITLE
bugs with closers wrapping around comments

### ISSUE THREAD

**Issue by oakmac** [oakmac/standard-clojure-style-js#60]
WIP with [PR-58](https://github.com/oakmac/standard-clojure-style-js/pull/58)

**Comment by oakmac** [comment]
This was a doozy of a problem to figure out. Hopefully [PR-58](https://github.com/oakmac/standard-clojure-style-js/pull/58) closes this out for good.

**Comment by oakmac** [comment]
> I'm following this project with great interest.

Thank you! Your interest is appreciated :nerd_face:  

> Why is the following considered beneficial?

The short answer is compatibility for [Parinfer] users. See [parinfer issue #92] for some historical context.

We also have the [gather trailing parens] recommendation from the Clojure Style Guide, which would apply in the following example:

```clj
[aaa
 [bbb ccc]
 [ddd eee]
 ;; fff
 ]
```

I will give this problem a think, but I am currently leaning towards the way it is written now.

I think the core issue comes down to "should comment lines be considered elements of the form they are in?", and I think the answer should be "no, they are not". There is probably some friction between the process of structural editing (ie: paredit) and that approach.

Undoubtedly, how to treat comments will be one of the harder parts of this project. And it is unlikely that the solution will be optimal for all possible code examples.

[Parinfer]:https://shaunlebron.github.io/parinfer/
[parinfer issue #92]:https://github.com/parinfer/parinfer.js/issues/92
[gather trailing parens]:https://guide.clojure.style/#gather-trailing-parens

**Comment by imrekoszo** [comment]
Thanks for taking the time to respond!

> compatibility for [Parinfer](https://shaunlebron.github.io/parinfer/) users

I can't say I'm a fan of parinfer - it is truly an impressive project which can make it easier for some folks to work with a LISP but I'm on the side of being in full control of the source code for which I find paredit and its derivatives perfect. I do understand your goal of making this tool play nice with all ways of editing code though.

> how to treat comments will be one of the harder parts of this project

I fully agree. Even paredit-likes don't always do what I expect them to do with comments. On the other hand, paredit doesn't prevent me from updating the code after to place comments where I think they should be.

> the [gather trailing parens](https://guide.clojure.style/#gather-trailing-parens) recommendation from the Clojure Style Guide

There's [another recommendation about margin comments](https://guide.clojure.style/#one-semicolon-for-margin-comments) there that is perhaps even more closely related and it also says "Avoid using those in situations that would result in hanging closing parentheses". I generally agree with this but I think there are exceptions.

> See https://github.com/parinfer/parinfer.js/issues/92

I'd be interested to know why you think _not_ doing this would break compatibility with Parinfer. I see there is some back-and-forth in the linked issue and even Shaun himself agrees that this is something controversial that Paredit does and created https://github.com/parinfer/parstager/issues/1 to perhaps mitigate the effect.

In my opinion - due to the reasons Shaun laid out - paredit is more or less forced to make a controversial source code change. Do you think this tool has to necessarily replicate that same controversial edit? My points are:

1. If a dev uses paredit and as a result makes this change, it won't be reverted by this tool, with or without #58 (if I understand correctly)
   - this means that if the dev triggering the change agrees with it, they can still have it
   - important to note that the code change here is triggered by an editing tool that the dev chose
2. A dev not using parinfer will be forced to suffer the change with #58
   - compared to point 1, the dev did not make the formatting change here
3. While I also prefer to gather trailing parens most of the time and in general that's a good advice I think being able to add comments this way is somewhat akin to whether you align map values or not - sometimes it's better to align and sometimes it's better not to. I think it's a wise choice from this tool to leave map value alignments up to the developer and my preference would be the same with regard to these trailing comments too

> "should comment lines be considered elements of the form they are in?" ... "no, they are not"

I agree with this in general, but even if they aren't elements of _forms_, they are a very important part of the source code itself and often extremely helpful when done well.

Just food for thought. Thanks for reading.

**Comment by oakmac** [comment]
I have given this much [hammock time] over the last several days and I think the best path forward is to wrap the closing parens "over" the comment. ie: how it is currently implemented via [PR-58]

[hammock time]:https://github.com/matthiasn/talk-transcripts/blob/master/Hickey_Rich/HammockDrivenDev.md
[PR-58]:https://github.com/oakmac/standard-clojure-style-js/pull/58

#### both ways are fine

If I were doing a code review and I saw either way of formatting, I don't think I would ever suggest changing to "the other way". In other words, I do not think the difference between these two styles has a large effect on the readability of the program: both are acceptable.

Given that, I think it makes sense to pick the way that can always be enforced (wrapping), does not cause friction with Parinfer users, and aligns with the Clojure Style Guide recommendation.

#### it is better to be consistent

To the extent that Standard Clojure Style can always produce the same result, it benefits everyone. I have a strong bias toward "making a decision" instead of "let the user choose".

This will not always be possible (ie: this project is not a strict pretty printer), and there are cases where I think it is important to allow the user to nudge the formatter in a direction ("Rule 3" indentation, vertically aligning `let` forms and maps), but every instance of flexibility allowed is a potential additional diff when reviewing code, which lessens the overall value of adopting this tool.

#### once it is "one way", I don't think this will be important

Colin Fleming has a [good quote] from the Clojureverse thread that spawned this project:

> "Formatting is something that devs tend to think is very important, until they can't control it and then in general as long as the output is decent they don't care."

The comment from user oskarth on [this issue] is instructive to consider: is the issue here the formatting? Or his expectation of what the formatting should be? I think it is the latter, and if tooling always made it work in the same way then that expectation can go away for everyone. This sort of comment on a Pull Request is _exactly_ the kind of small friction that I hope for Standard Clojure Style to remove for teams who adopt it.

[good quote]:https://clojureverse.org/t/clj-commons-building-a-formatter-like-gofmt-for-clojure/3240/25
[this issue]:https://github.com/parinfer/parinfer.js/issues/92#issuecomment-322785769

<hr>

Thank you for the interest and thoughtful comments @imrekoszo! If everyone who engages with this project brings the same level of thought and consideration as you, we are sure to all find success :blush:

**Comment by yuhan0** [comment]
Just for clarification – it wasn't entirely clear what "wrapping over" meant, and I had to inspect the commits in the linked PR to figure out what it was implementing.

Was the final decision to follow Parinfer's current behaviour? i.e. 

<table>
<tr><td>Original</td><td>After formatting</td></tr>
<tr><td><pre>
[aaa
 [bbb ccc]
 [ddd eee]
 ;; fff
 ]
</pre></td>
<td><pre>
[aaa
 [bbb ccc]
 [ddd eee]]
;; fff

</pre></td></tr>
</table>


I'm surprised that no one has mentioned the `#_` discard form, which is intended precisely for this sort of 'commenting out' of code. 
It's compatible with non-dangling parens and sidesteps the issue entirely, for the case of commented-out code:

<table>
<tr><td>Original</td><td>After formatting</td></tr>
<tr><td><pre>
[aaa
 [bbb ccc]
 [ddd eee]
 #_fff
 ]
</pre></td>
<td><pre>
[aaa
 [bbb ccc]
 [ddd eee]
 #_fff]

</pre></td></tr>
</table>

---

However this does not address the case of _non-code_ comments, which are the main topic of discussion in the ongoing linked Parinfer thread.

In standard lisp jargon, the proposed rule could be seen as the "barfing" of the line comment out into the outermost structure containing the first subsequent non-commented line - possibly top-level. (I suggest using this term instead of the potentially ambiguous "wrapping")

Since the implied nesting of comments within the code might convey significant information, having a one-size-fits-all formatter violate this structure doesn't seem ideal.

Here is a contrived example to illustrate such a risk:

<table>
<tr><td>Original</td><td>After formatting</td></tr>
<tr><td><pre lang="clj">
[:attributes
 [:size #{:small
          :medium
          :large}]
 [:color #{:red
           :green
           ;; WARNING: do not extend this list!
           }]
 [:other :prompt/user]]
</pre></td>
<td><pre lang="clj">
[:attributes
 [:size #{:small
          :medium
          :large}]
 [:color #{:red
           :green}]
 ;; WARNING: do not extend this list!
 [:other :prompt/user]]
 
</pre></td></tr>
<tr><td>Interpretation: no more colors allowed!</td>
<td> => *no more attributes allowed! </td></tr>
</table>

From the reader's perspective, this is of course a semantically invalid transformation. 
Even though there are obvious fixes like rephrasing the comment or placing it above the `[:color ..` line, these rely on spotting the "error" in the first place, which can be easy to miss in the context of an automatic formatter producing plenty of other semantically neglible diffs in the background.

**Comment by yuhan0** [comment]
### Alternative proposal 

One surprisingly robust workaround I've found over the years for this class of issues is to use a single comma character `,` as a sort of "anchor" for trailing delimiters - e.g. for the above examples:

```clj
[aaa
 [bbb ccc]
 [ddd eee]
 ;; fff
 ,]

(aaa
 [bbb
  [ccc
   ddd ;; need to use ddd here because of JIRA-1234
   ;; append more items to increase blergh threshold as needed
   ,] 
  ggg])

[:attributes
 [:size #{:small
          :medium
          :large}]
 [:color #{:red
           :green
           ;; WARNING: do not extend this list futher!
           ,}]
 [:misc :prompt/user]]
 ```

(Note: commas are syntactic whitespace in Clojure, so this has zero effect on the expression being read)

This happens to play nicely with language-agnostic Parinfer, which treats the comma as a symbol constituent, leaving it and any following delimiters untouched.

I don't recall where I first learnt this 'trick', but it seems to be relatively rare / undocumented in the wild.

Perhaps this should be brought up on the Clojure style guide repo - if the guide ends up recommending such a usage, ideally the formatter could detect such a situation and automatically insert these anchors:

<table>
<tr><td>Original</td><td>After formatting</td></tr>
<tr><td><pre lang=clojure>
[foo bar
 ]
</pre></td>
<td><pre lang=clojure>
[foo bar]

</pre></td></tr>
<tr><td><pre lang=clojure>
[foo ;; bar
 ]
</pre></td>
<td><pre lang=clojure>
[foo ;; bar
 ,]
</pre></td></tr>
<tr><td><pre lang=clojure>
[foo
 bar
 ;; TODO: implement quux
 ]
</pre></td>
<td><pre lang=clojure>
[foo
 bar
 ;; TODO: implement quux
 ,]
</pre></td></tr>
</table>

**Comment by imrekoszo** [comment]
@yuhan0 thanks for the alternative suggestion. Sadly, the important point you raise in your previous comment still applies:

> these rely on spotting the "error" in the first place

i.e. one needs to be aware that the formatter will make or has made a "semantically invalid transformation" and prepare for it/work around it

@shaunlebron's [suggested resolution](https://github.com/parinfer/parinfer.js/issues/92#issuecomment-2334830694) of the Parinfer issue appears very well thought out to me.

Allowing these still go against the general recommendation about [margin comments](https://guide.clojure.style/#one-semicolon-for-margin-comments) in the Style Guide, but as @yuhan0 points out, there will be exceptions when you want to be able to add comments like this.

**Comment by imrekoszo** [comment]
@yuhan0 just reacting to the examples you added in your [earlier comment](https://github.com/oakmac/standard-clojure-style-js/issues/60#issuecomment-2337840460):

I think cases B, C, and D should be left alone by formatters as there might be a semantic importance of the comments being inside the vector.

**Comment by yuhan0** [comment]
Yeah, that was my line of thought originally, but then I find it hard to imagine any scenario where one assigns semantic importance to the nesting of *inline* comments. 

Writing such comments 'in the margin' is nearly always an indication that they are directly referring to the code on that line, rather than belonging in any larger block structure.

In practice a far more common scenario is the following: an edit is made that leaves dangling parens following an inline comment: eg. deleting the forms `z (quz y)` in:
```clj
(let [x (foo) ;; Warning: side effects!
      y (bar x) ;; returns nil if invalid
      z (qux y)]
  (do-something x y))
```

Which results in this situation, where users may find difficulty getting their editor to un-dangle the `]` without (temporarily) breaking the structure.
In other words, a great use-case for letting a formatter deal with it automatically :) 

<table>
<tr><td><pre lang=clojure>
(let [x (foo) ;; Warning: side effects!
      y (bar x) ;; returns nil if invalid
      ]
  (do-something x y))
</pre></td>
<td><pre lang=clojure>
(let [x (foo) ;; Warning: side effects!
      y (bar x)] ;; returns nil if invalid
  (do-something x y))
</pre></td></tr>
</table>

**Comment by imrekoszo** [comment]
@yuhan0 I do find your `classify-squares` example to be okay, but I still wouldn't enfore the barf. The `let` and `comment` examples I find much better with the dangling closer.

> has to focus on being minimally invasive and respect one's intent

I think preserving intent is important no matter when the tool runs. It might even be more important for a check-in-time tool as you might not notice the tiny-yet-intent-hiding edit it makes when all you're thinking about is to finish the coding session.

> simply the fact that decades-old norms around formatting Lisp code very strongly proscribe against them

Wait till you learn that this tool throws out the decades-old Lisp tradition of [semantic indentation](https://metaredux.com/posts/2020/12/06/semantic-clojure-formatting.html) in favour of [fixed indentation](https://tonsky.me/blog/clojurefmt/) 😅

**Comment by yuhan0** [comment]
Yep, I'm aware of the indentation matter and I like the direction the tool is going there. My point was that such a deviation from tradition shouldn't be taken lightly – in that case the tradeoffs of REPL-driven vs static analysis were judged to be worth the cost of breaking tradition, a consideration that doesn't apply here.

In any case I'm not claiming to be the best judge of user intent / formatting choices etc, but I'm curious @oakmac what procedure you follow in such situations where there might not be a clear consensus on the Right Thing to do? (i.e. not covered by the existing Clojure Style guide either) 
Maybe collate them and have some sort of community poll near the end? (It's unclear what extent this is intended to be a personal vs community oriented project) Thanks for working on it!

**Comment by shaunlebron** [comment]
> Wait till you learn that this tool throws out the decades-old Lisp tradition of [semantic indentation](https://metaredux.com/posts/2020/12/06/semantic-clojure-formatting.html) in favour of [fixed indentation](https://tonsky.me/blog/clojurefmt/) 😅

@imrekoszo This isn’t true, right?  It compromises the two approaches by letting the user set the “semantic” indentation of the first line, then it “fixes” the rest of the lines to follow it.

**Comment by oakmac** [comment]
Just a quick note to say that I am monitoring the discussion in this thread and think there are some great points being raised. I am confident that the current behavior (as of v0.3.0) will need to change.

**Comment by oakmac** [comment]
From @yuhan0 [comment](https://github.com/oakmac/standard-clojure-style-js/issues/60#issuecomment-2338990379):

<table>
<tr><td>After edit:</td><td>Desired format:</td></tr>
<tr><td><pre lang=clojure>
(let [x (foo) ;; Warning: side effects!
      y (bar x) ;; returns nil if invalid
      ]
  (do-something x y))
</pre></td>
<td><pre lang=clojure>
(let [x (foo) ;; Warning: side effects!
      y (bar x)] ;; returns nil if invalid
  (do-something x y))
</pre></td></tr>
</table>

I agree with "Desired format" here.

If I saw the "After edit" format on a code review, I would probably suggest changing it to "Desired format", and that suggestion would be a nit, which is exactly the sort of PR comment that I hope Standard Clojure Style helps people avoid having to make.

**Comment by oakmac** [comment]
> One non-contrived situation where I suspect this rule will cause annoyance/friction is with "rich comment blocks"

I agree with this and think Standard Clojure Style needs a good answer / solution here.

I really like the `,))` trick because it 1) explicitly expresses developer intention to "hold" the parens and 2) plays nice with Parinfer.

I do not want Standard Clojure Style to add the comma automatically, but it may be necessary to promote "if you really do not want Standard Clojure Style to gather your trailing parens to one line, use a comma to hold them" advice.

**Comment by oakmac** [comment]
I should clarify that Standard Clojure Style is not being designed as a "tool for Parinfer users" anymore than it is a tool for Paredit users or users who use no editor tooling to help manage their parens (yes - there are people who work in this way).

It needs to support as many Clojure developers as it can, work in as many environments as it can, and reduce friction and diffs between users who edit their code in different ways.

In this particular use case, there is friction between Parinfer users and Paredit users because Parinfer users never have the condition of trailing parens below a single comment line, whereas a Paredit user might actively choose the code to look that way. It is not "one group thinks it should look one way, and the other way is wrong". It is "one group doesn't think about this at all, and the other group sometimes does".

I think that the idea of a `;;` comment being treated as "part of the structure" comes more from the editing process (ie: Paredit) than from reading the code. [`classify-squares`] is a great example of this IMO.

[`classify-squares`]:https://github.com/oakmac/standard-clojure-style-js/issues/60#issuecomment-2340644974

**Comment by shaunlebron** [comment]
The rules for margin comments in the Clojure Style Guide seemed off to me, so I’m proposing a clarification here: https://github.com/bbatsov/clojure-style-guide/issues/258

**Comment by shaunlebron** [comment]
> I really like the `,))` trick because it 1) explicitly expresses developer intention to "hold" the parens and 2) plays nice with Parinfer.
> 
> I do not want Standard Clojure Style to add the comma automatically, but it may be necessary to promote "if you really do not want Standard Clojure Style to gather your trailing parens to one line, use a comma to hold them" advice.

Ha, I like this trick and completely forgot about it.

One thing to consider is that if the formatter requires the comma to be there, but doesn’t add the commas automatically, then it will create a lot of initial diffs in existing codebases have to be manually corrected per project.

Assuming I understand this correctly, if we want to avoid those diffs, this means either 1) Standard Clojure Style has to add them in automatically, or 2) Parinfer has to be updated to not require them.

**Comment by imrekoszo** [comment]
> > Wait till you learn that this tool throws out the decades-old Lisp tradition of [semantic indentation](https://metaredux.com/posts/2020/12/06/semantic-clojure-formatting.html) in favour of [fixed indentation](https://tonsky.me/blog/clojurefmt/) 😅
> 
> @imrekoszo This isn’t true, right? It compromises the two approaches by letting the user set the “semantic” indentation of the first line, then it “fixes” the rest of the lines to follow it.

@shaunlebron by fixed indentation I meant "Multi-line lists that start with a symbol are always indented with two spaces" per the Nikita's article, with the addition of your rule "3. multi-line lists are indented to "first-arg alignment" if and only if the 2nd line is indented as such". (It would perhaps be useful to give a name to this composite style, perhaps Soft Fixed or Hintable Fixed indentation 😅 ) 

By semantic indentation I meant what's in Bozhidar's article, with further details in the [CIDER Indentation](https://docs.cider.mx/cider/config/indentation.html), and [Indentation Specification](https://docs.cider.mx/cider/config/indentation.html) articles. The bottom line of these is that formatting needs to understand the semantics of forms and apply different formatting rules based on those semantics (special args indented one space, body indented two spaces, etc.)

In the fixed vs semantic debate I agree with the Hintable Fixed position due to its general applicability (formatter doesn't need to know the semantics), but still letting the programmer (who does know) to align with the first arg if they consider that more expressive.

On the other hand, this allowing of Rule 3 carries the potential of PR "nit" comments, similar to what @oakmac [points out](https://github.com/oakmac/standard-clojure-style-js/issues/60#issuecomment-2342301466) above. Both the following are valid according to Standard Clojure Style, but only one each per either Semantic of Strict Fixed.

```clj
;; semantic ❌ strict fixed ✅ hintable fixed ✅
(map +
  (range 10)
  (range 10))

;; semantic ✅ strict fixed ❌ hintable fixed ✅
(map +
     (range 10)
     (range 10))
```

For the sake of completeness, here are the narrow formats too, of which each of the 3 formatting specifications only allow one:

```clj
;; semantic ✅ strict fixed ❌ hintable fixed ❌
(map
 +
 (range 10)
 (range 10))

;; semantic ❌ strict fixed ✅ hintable fixed ✅
(map
  +
  (range 10)
  (range 10))
```

I suppose there is a symmetry between "requiring a comma to prevent barfing the comment" and "requiring (manual/tool-assisted) lining up the second arg with the first to prevent the strict 2-space indent". Both of these require effort from the programmer in case they expect a benefit from the "alternative" formatting.

While aligning with the first arg is common practice, leaving commas around is less so, and Shaun is right to point out the possibility of additional initial diffs. Note that since the narrow indentation format is changing from semantic to fixed, that will probably bring even more initial diffs. However, those diffs in my opinion preserve programmer intent better than barfing a trailing margin comment, but again, it could be just me.

**Comment by oakmac** [comment]
>  if the formatter requires the comma to be there, but doesn’t add the commas automatically, then it will create a lot of initial diffs in existing codebases that have to be manually corrected per project

It is not a goal of Standard Clojure Style to "avoid initial diffs when applied to a codebase for the first time". I suspect that for 99% of forms, gathering the trailing parens to one line generally improves the readability of the code. The "comment line holding trailing parens" is the exception, not the norm.

**Comment by imrekoszo** [comment]
> "Standard Clojure Style will always gather trailing parens to a single line (as recommended by the Clojure Style Guide and enforced by Parinfer). If you want a `;` comment line to 'hold' some parens on the next line, place a comma before them to explicitly show that intent."

@oakmac I think it's still beneficial to allow indented comments to preserve their position in the AST, per https://github.com/parinfer/parinfer.js/issues/92#issuecomment-2334830694

**Comment by shaunlebron** [comment]
> "Standard Clojure Style will always gather trailing parens to a single line (as recommended by the Clojure Style Guide and enforced by Parinfer). If you want a `;` comment line to 'hold' some parens on the next line, place a comma before them to explicitly show that intent."

I was thinking about changing the Parinfer enforcement, but maybe that ship has sailed.  I’m at least leaning toward an option for auto-inserting the commas.

I requested another clarification in the Clojure Style Guide to see if we can add a recommendation there about indented comments, or at least a mention.

**Comment by oakmac** [comment]
Relevant to this thread, I learned a fun way to "hold" parens on the next line from @aiba at Clojure/conj 2024. Example rich comment block:

```clj
(comment
  (something)
  (something-else :foo :bar)
  (another-thing)
  :-)
```
